### PR TITLE
Test environment for local Openwhisk Java functions

### PR DESCRIPTION
**Pull request by gorkem**
This is the Initial implementation for a CLI and runtime for testing Java functions locally. It functions very similar to node-local.

### CODE REVIEW

**Comment by dubeejw** [comment]
Hey @gorkem, I remember you from the thym repo!

File: java-local/src/main/java/openwhisk/java/local/Launcher.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.Map;
+
+import com.google.gson.JsonObject;
+
+/**
+ * Local launcher for openwhisk functions
+ *
+ * @author Gorkem Ercan
```
Comment (csantanapr) [inline java-local/src/main/java/openwhisk/java/local/Launcher.java]: Could you remove the `@author` line? I don't want the Apache police after us

File: java-local/src/main/java/openwhisk/java/local/MemoryFileManager.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import javax.tools.FileObject;
+import javax.tools.ForwardingJavaFileManager;
+import javax.tools.JavaFileManager;
+import javax.tools.JavaFileObject;
+import javax.tools.SimpleJavaFileObject;
+
+/**
+ * Keeps compiled class in memory
+ *
+ * @author Gorkem Ercan
```
Comment (csantanapr) [inline java-local/src/main/java/openwhisk/java/local/MemoryFileManager.java]: Could you remove the @author line? I don't want the Apache police after us

File: java-local/src/main/java/openwhisk/java/local/CLI.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+import picocli.CommandLine;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.ParameterException;
+import picocli.CommandLine.Parameters;
+
+/**
+ * CLI
+ *
+ * @author Gorkem Ercan
```
Comment (csantanapr) [inline java-local/src/main/java/openwhisk/java/local/CLI.java]: Could you remove the @author line? I don't want the Apache police after us

File: java-local/src/main/java/openwhisk/java/local/JarLoader.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import java.lang.reflect.Modifier;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Path;
+
+/**
+ * Parts copied from https://github.com/apache/incubator-openwhisk-runtime-java/blob/master/core/javaAction/proxy/src/main/java/openwhisk/java/action/JarLoader.java
+ *
+ * @author Gorkem Ercan
```
Comment (csantanapr) [inline java-local/src/main/java/openwhisk/java/local/JarLoader.java]: Could you remove the @author line? I don't want the Apache police after us

**Comment by csantanapr** [comment]
@rabbah can you give it a review?

**Comment by gorkem** [comment]
@csantanapr removed the author annotations. Also cleaned up a couple of console outputs that did not make much sense.

**Comment by csantanapr** [comment]
@gorkem I'm having trouble running this

I created the jar with `./gradlew jar`
Then running
```
java -jar ./build/libs/java-local.jar myproject/java-1.0.jar --main=Hello
```
It hangs it never returns.

The `java-1.0.jar` is simple one
```java
import com.google.gson.JsonObject;

public class Hello {
    public static JsonObject main(JsonObject args) {
        String name = "stranger";
        if (args.has("name"))
            name = args.getAsJsonPrimitive("name").getAsString();
        JsonObject response = new JsonObject();
        response.addProperty("greeting", "Hello " + name + "!");
        return response;
    }
}
```

ping me on Slack

**Comment by csantanapr** [comment]
@gorkem ping

**Comment by gorkem** [comment]
@csantanapr Updated the PR with some additional testing and also the timeout issue should be gone.

**Comment by dubeejw** [comment]
@gorkem, Travis failed. I restarted it just to verify if the failure happens again.

**Comment by gorkem** [comment]
@dubeejw I think that was because master was already broken. Seems to be fixed can you kick the build again?

**Comment by csantanapr** [comment]
I just saw this, Thanks @gorkem 
It works now for me
```
🐑  $ java -jar ./build/libs/java-local.jar  myproject/java-1.0.jar --main=Hello name=Carlos
{"greeting":"Hello Carlos!"}

🐑  $ java -jar ./build/libs/java-local.jar  myproject/java-1.0.jar --main=Hello
{"greeting":"Hello stranger!"}
```

**Pull request merged**